webui/src/lib/hooks/router.ts

```diff
@@ -0,0 +1,52 @@
+import {
+  useHistory,
+  useLocation,
+  useParams,
+  generatePath,
+} from "react-router-dom";
+
+export const useQuery = <T extends Partial<T>>(): T => {
```

Comment:
Also note that this function lies about its return type: it never checks that has all keys for T.  So it's really
```suggestion
export const useQuery = <T>(): Partial<T> => {
```

Reply:
Good point! I agree that forcing the consumer to "`undefined` check" the result is a good thing.